### PR TITLE
[POC] support window/over function

### PR DESCRIPTION
**Pull request by walterddr**

### CODE REVIEW

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryEnvironmentTestBase.java [hunk 1]
```diff
@@ -67,6 +67,7 @@ protected Object[][] provideQueries() {
         new Object[]{"SELECT dateTrunc('DAY', a.ts + b.ts) FROM a JOIN b on a.col1 = b.col1 AND a.col2 = b.col2"},
         new Object[]{"SELECT a.col2, a.col3 FROM a JOIN b ON a.col1 = b.col1 "
             + " WHERE a.col3 >= 0 GROUP BY a.col2, a.col3"},
+        new Object[]{"SELECT a.col1, SUM(a.col3) OVER (PARTITION BY a.col2) FROM a"},
```
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/QueryEnvironmentTestBase.java:70]: yes this PR is for demo propose.. I will create a complete one soon for planner and for runtime.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  public WindowNode(int stageId, List<Window.Group> windowGroups, List<RexLiteral> constants,
+      DataSchema dataSchema) {
+    super(stageId, dataSchema);
+    _windows = new ArrayList<>(windowGroups.size());
+    for (Window.Group group : windowGroups) {
+      _windows.add(new PinotWindow(group.keys, group.orderKeys, group.aggCalls, group.lowerBound, group.upperBound));
+    }
+  }
+
+  public static class PinotWindow {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java:55]: WindowInfo ?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+public class WindowNode extends AbstractStageNode {
+  @ProtoProperties
+  private List<PinotWindow> _windows;
+
+  public WindowNode(int stageId) {
+    super(stageId);
+  }
+
+  public WindowNode(int stageId, List<Window.Group> windowGroups, List<RexLiteral> constants,
```
Comment (amrishlal) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java:46]: Does WindowNode represent a single window function expression in the select list or does it represent multiple window function expressions? For example, what would WindowNode look like for the following sql:
`SELECT lname, avg(score) over(), avg(score) over(partition by lname) from mytable`?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      DataSchema dataSchema) {
+    super(stageId, dataSchema);
+    _windows = new ArrayList<>(windowGroups.size());
+    for (Window.Group group : windowGroups) {
+      _windows.add(new PinotWindow(group.keys, group.orderKeys, group.aggCalls, group.lowerBound, group.upperBound));
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java:51]: So each `Window.Group` represents a set of window functions with the same PARTITION BY, ORDER BY and frame / bound characteristics ?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  public static class PinotWindow {
+    public List<RexExpression> _groupSet;
+    public List<RexExpression> _orderSet;
+    public List<RexExpression> _aggCalls;
+    public int _lowerBound;
+    public int _upperBound;
+
+    public PinotWindow(ImmutableBitSet keys, RelCollation orderKeys, ImmutableList<Window.RexWinAggCall> winAggCalls,
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java:62]: So, ideally this should handle all kinds of Window functions and not just Aggregate class of functions. For example, LAG, LEAD, NTILE etc will also be added at some point.

Does `RexWinAggCall` capture everything ?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RelToStageConverter.java [hunk 6]
```diff
@@ -77,11 +79,17 @@ public static StageNode toStageNode(RelNode node, int currentStageId) {
       return convertLogicalAggregate((LogicalAggregate) node, currentStageId);
     } else if (node instanceof LogicalSort) {
       return convertLogicalSort((LogicalSort) node, currentStageId);
+    } else if (node instanceof LogicalWindow) {
```
Comment (amrishlal) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RelToStageConverter.java:82]: I ran the following query and tried to see what `relRoot` in `StagePlanner.makePlan` look like.
`select playerID, avg(runs) over(partition by playerID) from baseballStats limit 10 OPTION(useMultistageEngine=true)`
For this query, relRoot has:
```
LogicalSort
  LogicalProject
     exps (CAST(/(SUM($26) OVER (PARTITION BY $27), COUNT($26) OVER (PARTITION BY $27))):INTEGER NOT NULL)
     LogicalTableScan
```
`LogicalWindow` does not seem to appear in the relRoot. Also avg(runs) over(partition by playerID) got replaced by cast, sum, and count functions? I am wondering if we can avoid such rewrites?

**Comment by walterddr** [comment]
closing as plan changed

**Pull request closed without merging**